### PR TITLE
Consider children Control nodes for mouse-enter/exit notifications

### LINKED ISSUE #81909: MouseEnter/Exit events not correctly passed on from children to parent script since Godot 4.2 dev3
### Godot version

4.2 dev3 and 4.2 dev4

### System information

Windows 10

### Issue description

With the following Control Node hierarchy:
image
the `TextureRect` Node (godot icon) blocks the `mouse_entered / mouse_exited` signals in the script of the `ParentWithScript` Node (the white area). The `TextureRect` uses its default `mouse_filter` of `Pass`. When using the `mouse_filter` `Ignore`, the signals are triggered correctly on the parent.

I went back the official versions to see when the behavior started and it seems to have started with 4.2 dev3.

In 4.2 dev2 the behavior is still a bit weird, but it has been like that for a long time (tested with 4.1.1) : when moving the mouse from "over the parent" (the white area in the MRP) directly to "over the child" (the godot icon), the `mouse_exited` signal gets triggered first and then directly the `mouse_entered` again. But at least the last signal that gets triggered is `mouse_entered` (important for custom tooltips!).

Another quick observation (that is not in the MRP): I added a tooltip to the `ParentWithScript` node and that gets correctly shown when hovering over the `TextureRect` child node! So it seems to only affect the signals.

### Steps to reproduce

- Open the minimal reproduction project in Godot 4.2 dev3 or dev4
- Run it
- Observe that `entered` gets printed when hovering over the white area (the parent), but `exited` gets called as soon as the mouse moves over the godot icon (the child)

### Minimal reproduction project

[MouseEnterExitTest.zip](https://github.com/godotengine/godot/files/12657936/MouseEnterExitTest.zip)

### LINKED ISSUE #82530: Problems with new `mouse_entered` and `mouse_exited` signals
### Godot version

master, 4.2 dev3+, since 1c3c17

### System information

Windows 10

### Issue description

#81909 is caused by the same root cause, #67791
#82182 does not solve the issues.

### Mouse enter and exit are not in sync with _gui_input InputEventMouseButton and other mouse events.
This leads to issues where a button can be clicked on without being hovered:
godot_mouse_event_issue
InputEventMouseMotion should be called only between mouse enter and exit events, currently this is not the case.
The existing mouse events (InputEventMouse*) are fine and work as expected, the mouse enter and exit signals need to match it.

### Mouse enter and exit do not respect `MOUSE_FILTER_PASS`.
Only `MOUSE_FILTER_IGNORE` and `MOUSE_FILTER_STOP` are respected. This means there is not a good way to get the event in a parent while the child needs mouse events.
Workarounds include: 
- Manually set all child Controls to `MOUSE_FILTER_IGNORE`. This does not scale and doesn't work if the children also need mouse input. (Doesn't `MOUSE_FILTER_PASS` exist for this reason?)
- Manually track the proper state by listening to InputEventMouseMotion. This can tell you when the mouse has entered, but not exited.
- Continuously check `Rect2.has_point()`. This is only axis-aligned and doesn't include children.

These workarounds are cumbersome and not performant.
All Nodes that use `NOTIFICATION_MOUSE_ENTER` like `BaseButton` would also need to do be changed.

The current implementation of mouse enter and exit is useful in its own way, but should not be the default used for most events.

### No other UI system I could find does mouse enter and exit events like this. 
For the ones that provide multiple ways, this way isn't the default.
- [Unity has both mouse enter/leave and mouse over/out](https://docs.unity3d.com/Manual/UIE-Mouse-Events.html)
- [HTML has both onmouseenter/onmouseleave (doesn't bubble up, excludes overlaps, includes children) and onmouseover/onmouseout (bubbles up, excludes overlaps, includes children)](https://www.w3schools.com/jsref/event_onmouseenter.asp)
- [Windows has both IsMouseOver (includes child elements)](https://learn.microsoft.com/en-us/dotnet/api/system.windows.uielement.ismouseover) [and IsMouseDirectlyOver (does not)](https://learn.microsoft.com/en-us/dotnet/api/system.windows.uielement.ismousedirectlyover?view=windowsdesktop-7.0#system-windows-uielement-ismousedirectlyover)
- [Phaser and has a toponly input flag to ignore overlaps](https://newdocs.phaser.io/docs/3.60.0/Phaser.Input.InputPlugin#topOnly)
- [Haxeflixel has mouseChildren flag that allows children to receive mouse events](https://api.haxeflixel.com/flixel/input/mouse/FlxMouseEventManager.html)
- [Gamemaker does bounding box test, no overlaps](https://manual.yoyogames.com/#t=The_Asset_Editors%2FObject_Properties%2FObject_Events.htm&rhsearch=mouse%20enter&rhhlterm=mouse%20enter)

### Proposed Solution
Change the current `mouse_enter` and `mouse_exited` signals to `mouse_over` and `mouse_out`. Add a `NOTIFICATION_MOUSE_OVER`  and `NOTIFICATION_MOUSE_OUT` for it.
`mouse_entered` signal: Emitted when the mouse enters the control's Rect area or any of its descendants, provided its [mouse_filter](https://docs.godotengine.org/en/stable/classes/class_control.html#class-control-property-mouse-filter) lets the event reach it.
`mouse_exited` signal: Emitted when the mouse exits the control's Rect area and all of its descendants, provided its [mouse_filter](https://docs.godotengine.org/en/stable/classes/class_control.html#class-control-property-mouse-filter) lets the event reach it.
The signals should not propagate up if the parent already received that signal.
Implementation details for this:
- The mouse entered signal would propagate up until the last gui.mouse_over,
- The mouse exited signal would propagate up until where the new gui.mouse_over is a parent of the old gui.mouse_over.
- otherwise they do propagate like the mouse events do.

This ensures that a Control first receives `mouse_enter` then mouse move events then `mouse_exit`.

### Steps to reproduce

Add a button
Add a visible control to it with `MOUSE_FILTER_PASS` set, like an image.
Hover over the button and the image. 
The button does not get hovered, even though it will accept a mouse click.

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by Sauermann**
`NOTIFICATION_MOUSE_ENTER` and `NOTIFICATION_MOUSE_EXIT` and the according signals now includes the areas of children control nodes.

In order to check if a Control node itself was entered/exited, the newly introduced `NOTIFICATION_MOUSE_ENTER_SELF` and
`NOTIFICATION_MOUSE_EXIT_SELF` can be used.

resolve #81909
resolve #82530
regression from #67791
supersedes #82182

### CODE REVIEW

**Comment by Sauermann** [comment]
Thanks for testing and the valuable feedback, I appreciate it. This is still a draft, because it has some bugs. I will mark it ready for review, when I have addressed the bugs.

File: scene/main/viewport.cpp [hunk 1]
```diff
@@ -3070,15 +3074,107 @@ void Viewport::_update_mouse_over(Vector2 p_pos) {
...
+			Control *control = over;
+			while (control) {
+				control_list.push_back(control);
+
+				if (control->is_set_as_top_level()) {
+					// Top level Control nodes break the propagation chain.
+					break;
+				}
+
+				// Consider only direct Control parents, but not parents, that are separated by non-Control CanvasItems. (Could be changed to inclide them.)
```
[the remark this answers is not in the document]
Comment (Sauermann) [reply scene/main/viewport.cpp]: Thanks, I have corrected the typos.

**Comment by Sauermann** [comment]
I have redone the PR and updated my initial naive approach with a more robust method that stores in addition to `mouse_over` the interval of the nodes, that currently have received a `NOTIFICATION_MOUSE_ENTER` without having received a `NOTIFICATION_MOUSE_ENTER_SELF`.
This was necessary to be able to correctly handle the cleanup when removing hovered nodes, which wasn't possible in the initial version.

I have also tested the MRP in the linked issue and also verified, that all tests pass.
Nonetheless, I believe, that this PR would benefit from thorough testing, since it changes core behavior.
I am aware, that changing `top_level` of hovered Control nodes is currently not handled correctly.

**Comment by kitbdev** [comment]
It looks like mouse filters are completely ignored.
And when hiding or removing a node if the mouse is not above it's parent the parent doesn't get any exit notifications.

**Comment by kitbdev** [comment]
I put my take on this here if you want to check out: https://github.com/kitbdev/godot/commit/1eda4ce244323fa91971cb7dcb1132621cc22777
It's based off of your previous commit. 
It currently doesn't properly handle changing the mouse filter or top level while it or it's descendant is hovered.

**Comment by KoBeWi** [comment]
> It looks like mouse filters are completely ignored.

IGNORE seems to work correctly, STOP and PASS do the same though.

> And when hiding or removing a node the parent doesn't get any exit notifications.

Can't reproduce, the node correctly receives mouse exit when hidden. If shown under cursor, the entered notification is not received until mouse movement, but I assume this is expected for performance reasons.

**Comment by WolfgangSenff** [comment]
Just checking because I added some comments to various, seemingly-related but already-closed bugs (https://github.com/godotengine/godot/issues/17326), but would this solve an issue where I'm trying to drag-and-drop a control into a subviewport on top of another control in said SVP? Seems like it should, but I'm not as familiar with this exact part of the codebase.

**Comment by Sauermann** [comment]
@WolfgangSenff Dragging into different SubViewports is currently not supported. #67531 intends to establish Drag&Drop support for differnet SubViewports. This PR should not have any influence on Drag&Drop.

**Comment by akien-mga** [comment]
Superseded by #84547.

**Pull request closed without merging**